### ISSUE TITLE
TFException: Attempting to use uninitialized value

### ISSUE THREAD

**Issue by ohernpaul** [migueldeicaza/TensorFlowSharp#251]
Hello,

I'm attempting to import a model created in Keras/Tensorflow and use it for inference in a Unity project.

I have successfully imported the model and validated by printing names of input and output nodes in the graph. Though, when I try to get the output value from the runner, I get this exception.

TFException: Attempting to use uninitialized value action_b
	 [[Node: action_b/read = Identity[T=DT_FLOAT, _class=["loc:@action_b"], _device="/job:localhost/replica:0/task:0/cpu:0"](action_b)]]
TensorFlow.TFStatus.CheckMaybeRaise (TensorFlow.TFStatus incomingStatus, System.Boolean last) (at <6ed6db22f8874deba74ffe3e566039be>:0)
TensorFlow.TFSession.Run (TensorFlow.TFOutput[] inputs, TensorFlow.TFTensor[] inputValues, TensorFlow.TFOutput[] outputs, TensorFlow.TFOperation[] targetOpers, TensorFlow.TFBuffer runMetadata, TensorFlow.TFBuffer runOptions, TensorFlow.TFStatus status) (at <6ed6db22f8874deba74ffe3e566039be>:0)
TensorFlow.TFSession+Runner.Run (TensorFlow.TFStatus status) (at <6ed6db22f8874deba74ffe3e566039be>:0)
RecordArbitraryData.AllInOne (System.Single[] input) (at Assets/Scripts/Spells/RecordArbitraryData.cs:324)
RecordArbitraryData.FixedUpdate () (at Assets/Scripts/Spells/RecordArbitraryData.cs:101)

Here is a sanity check function I used to attempt to make a prediction. The model is being imported correctly because I can see the input and output layer's names with debug. 
```
void AllInOne(float[] input){
		var graph = new TFGraph();
		string model_name = "simple_as_binary.pb";
		var model = File.ReadAllBytes(model_dir + model_name);
		Debug.Log(model_dir + model_name);
		graph.Import (model, "");

		session = new TFSession (graph);
		using (session) {
			//Assign input tensors
			var runner = session.GetRunner ();
			runner.AddInput (session.Graph [input_node_name] [0], input);

			//Calculate and access output of graph
			runner.Fetch (session.Graph [output_node_name] [0]);
			float[,] recurrent_tensor = runner.Run () [0].GetValue () as float[,];
		}
	}
```
@migueldeicaza  Do you have any idea what is happening? Is my graph being initialized incorrectly? Is there a step im missing after importing my model?

**Comment by migueldeicaza** [comment]
You can use this to debug your problem:

https://www.tensorflow.org/programmers_guide/graph_viz

**Comment by ohernpaul** [comment]
I was able to figure out most of my problems. I'm currently at a point where my model is predicting in unity, but only predicting the first of four classes. My guess is it has something to do with the weights not getting initialized correctly from the checkpoint files?

Preface: Mozilla Firefox works best for displaying tensorboard; it took me a long time to realize that google chrome was causing my graph to be invisible (tensorboard is how I was able to figure out the nodes that needed to be used for input and output).

First Issue: I was renaming a .pb file into a .bytes file. This is incorrect because the model’s weights come from the checkpoint file, and are given to the nodes held in the .pb file.

Second Issue: I was using the file created called ‘checkpoint’, which was throwing an error. I then changed the name of the checkpoint to ‘test’ and used this in the freeze_graph function. When calling the checkpoints file, I was required to use ‘test.ckpt’. I assume this function knows to grab the three files automatically based on the .ckpt? ‘Test’ without ‘.ckpt’  did not work. 

Third Issue: when using the freeze_graph function, I needed to export the .pb file in keras/tf with text=False. I tested True and False; True threw an error about “bad wiring”.

Fourth Issue: Tensorboard was very difficult to use without any organization. Using tf.name_scope helped a lot with not only visualization, but making sure I was using/referencing the correct nodes in TensorFlowSharp. In keras I found it helpful to separate the final Dense layer and Activation into their own scopes so I could find the correct output node. The rest of my network was put into a ‘body’ scope, and the sole input layer in ‘input’ scope. The name_scope function prepends ‘scopename/’ to the node name. I don't think it’s necessary, but it helped me.

Fifth Issue: The version of tensorflowsharp released as a unity package is not up to date. This caused an issue with a keras placeholder for ‘keras_training_phase’. In keras, you pass this along with an input like [0] + input. I tried to do the same by creating a new TFTensor(bool), but I was getting an error ‘inaccessible due to its protection level. This was an error with the implicit conversion between bool and TFTensor in my unity TensorFlowSharp version. To fix this I had to use a function found in this [stackoverflow solution](https://stackoverflow.com/questions/43332342/is-it-possible-to-replace-placeholder-with-a-constant-in-an-existing-graph/43342922) where the .bytes file is read in, the placeholder for keras_training_phase is found, and is swapped out for a bool constant set to false. This worked for me because my model was pretrained in python, so it may not be a great fix for someone that’s trying to train and test the model. A condition for removing this node with the freeze_graph function would really be useful.

Hope someone finds this useful!
